### ISSUE TITLE
Change the `path!` macro to assume `end()`

### ISSUE THREAD

**Issue by seanmonstar** [seanmonstar/warp#357]
The ability to chain together path prefixes is very powerful and explicit, but it can be a gotcha to forget to leave off a `warp::path::end()` filter. I suspect this is especially true when using the `path!` macro, since one is likely building the entire path then.

I'd like to suggest the `path!` macro change such that it will by default include a `warp::path::end()` filter unless the user explicitly specifies not to. This would be done by adding a special syntax to signal its a "prefix" path. My suggestion for that syntax is to add `/ ..` to the path, but I'm open to other syntax.

## Examples:

- `path!("foo" / ..)`: This just the prefix `"foo/"`.
- `path!("foo" / "bar" / ..)`: This is the prefix `foo/bar/".
- `path("foo" / "baz")`: This is the suffix `foo/baz`.
- `path("foo"/ .. / "bar")`: This is a compile error.

## Alternative

We could instead add special syntax to opt-**in** instead, something like `path!("foo" / "baz" $)`. but I suspect it's more common with `path!` to spell out the full path, and should just default to that.

**Comment by seanmonstar** [comment]
@jxs thoughts?

**Comment by tshepang** [comment]
I am fan of main proposal, especially since it has an opt-out (complete with nice syntax too).

**Comment by gperinazzo** [comment]
I think assuming `end()` is a good idea. When I first started building an API with warp, I had a few problems with forgetting to `end()` paths, and then having them match with sub-paths by mistake.

In the end, I found that always adding `end` when building a route was usually better, even if there's currently no sub-path. The [docs for `warp::path`](https://docs.rs/warp/0.1.20/warp/filters/path/index.html) could also be updated to include a section about it, as it's currently lacking any mention to it and no example uses it.
